File: express/app/src/main/java/com/changsukuaidi/www/base/BaseFragment.java
package com.changsukuaidi.www.base;

import android.databinding.DataBindingUtil;
import android.databinding.ViewDataBinding;
import android.graphics.BitmapFactory;
import android.os.Build;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;

import com.changsukuaidi.www.R;
import com.changsukuaidi.www.base.i.IBasePresenter;
import com.changsukuaidi.www.base.i.IBaseView;
import com.changsukuaidi.www.common.utils.DisplayUtils;
import com.changsukuaidi.www.databinding.ToolbarContentBinding;
import com.changsukuaidi.www.utils.ToastUtils;
import com.changsukuaidi.www.widget.popup.LoadingDialog;
import com.tbruyelle.rxpermissions2.RxPermissions;
import com.trello.rxlifecycle2.components.support.RxFragment;

import org.greenrobot.eventbus.EventBus;

import javax.inject.Inject;

/**
 * @Describe
 * @Author thc
 * @update zhouhao
 * @Date 2017/11/30
 * @Contact 605626708@qq.com
 */

public abstract class BaseFragment<P extends IBasePresenter, VBD extends ViewDataBinding> extends RxFragment implements IBaseView {
    @Inject
    protected P mPresenter;

    protected View rootView;

    protected ToolbarContentBinding mTitleViewBinding;
    protected VBD mViewBinding;

    protected RxPermissions mRxPermissions;
    protected LoadingDialog mLoadingDialog;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, Bundle savedInstanceState) {
        initIntent();

        mViewBinding = DataBindingUtil.inflate(inflater, layoutId(), container, false);

        // 添加title
        if (useTitle()) {
            LinearLayout linearLayout = new LinearLayout(getContext());
            linearLayout.setLayoutParams(new ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT));
            linearLayout.setOrientation(LinearLayout.VERTICAL);

            // 有标题的时候都给statusBar的背景设置为themeColor
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                getActivity().getWindow().setStatusBarColor(getResources().getColor(R.color.themeColor));
            }

            // title layout
            mTitleViewBinding = DataBindingUtil.inflate(inflater, R.layout.toolbar_content, container, false);
            setToolbarLeftEnable();

            // add title
            linearLayout.addView(mTitleViewBinding.getRoot());
            // add content
            linearLayout.addView(mViewBinding.getRoot(), new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT));
            rootView = linearLayout;
        } else {
            if (showFullScreen()) {
                // 设置全屏显示,占用statusBar背景位置
                getActivity().getWindow().getDecorView().setSystemUiVisibility(View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN);
            } else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                // 设置状态栏背景为themeColor
                getActivity().getWindow().setStatusBarColor(getResources().getColor(R.color.themeColor));
            }
            rootView = mViewBinding.getRoot();
        }

        setPresenter();
        if (useEventBus()) {
            EventBus.getDefault().register(this);
        }
        if (usePermissions()) {
            //noinspection ConstantConditions
            mRxPermissions = new RxPermissions(getActivity());
        }
        initView();
        initData();
        return rootView;
    }

    protected void initIntent() {

    }

    protected abstract void setPresenter();

    protected abstract int layoutId();

    protected void initView() {

    }

    protected void initData() {

    }

    @Override
    public void onDetach() {
        super.onDetach();
        if (mPresenter != null)
            mPresenter.onDestroy();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        if (useEventBus())
            EventBus.getDefault().unregister(this);
    }

    protected boolean onBackPressed() {
        return false;
    }

    protected boolean useEventBus() {
        return false;
    }

    protected boolean usePermissions() {
        return false;
    }

    @Override
    public void showSuccessMessage(String message) {
        ToastUtils.showToast(getContext(), message);
    }

    @Override
    public void showFailedMessage(String message) {
        ToastUtils.showToast(getContext(), message);
    }

    @Override
    public void showLoading(String msg) {
        if (mLoadingDialog == null) {
            mLoadingDialog = LoadingDialog.newInstance(msg);
        }
        mLoadingDialog.show(getFragmentManager(), "loading");
    }

    @Override
    public void showLoading() {
        if (mLoadingDialog == null) {
            mLoadingDialog = LoadingDialog.newInstance("请稍后...");
        }
        mLoadingDialog.show(getFragmentManager(), "loading");
    }

    @Override
    public void hideLoading() {
        mLoadingDialog.dismiss();
    }

    @Override
    public void onNetError() {

    }

    @Override
    public void showEmpty() {

    }

    /**
     * 确定是否需要全屏展示
     */
    public boolean showFullScreen() {
        return false;
    }

    public boolean useTitle() {
        return true;
    }

    protected void setToolbarTitle(String title) {
        mTitleViewBinding.toolbarTitle.setText(title);
    }

    protected void setToolbarTitleWithRightDrawable(String title, int resId, View.OnClickListener listener) {
        mTitleViewBinding.toolbarTitle.setText(title);
        mTitleViewBinding.toolbarTitle.setCompoundDrawablesWithIntrinsicBounds(0, 0, resId, 0);
        mTitleViewBinding.toolbarTitle.setCompoundDrawablePadding(DisplayUtils.dp2px(getContext(), 10));
        mTitleViewBinding.toolbarTitle.setOnClickListener(listener);
    }

    protected void setToolbarLeftEnable() {
        mTitleViewBinding.toolbarLeft.setVisibility(View.VISIBLE);
        mTitleViewBinding.toolbarLeft.setImageBitmap(BitmapFactory.decodeResource(getResources(), R.mipmap.icon_back));
        mTitleViewBinding.toolbarLeft.setOnClickListener(view -> getActivity().onBackPressed());
    }

    protected void setToolbarLeftDisEnable() {
        mTitleViewBinding.toolbarLeft.setVisibility(View.INVISIBLE);
    }

    protected void setToolbarRightInvisible() {
        mTitleViewBinding.toolbarRight.setVisibility(View.INVISIBLE);
    }

    protected void setToolbarRight(String text, View.OnClickListener listener) {
        mTitleViewBinding.toolbarRight.setVisibility(View.VISIBLE);
        mTitleViewBinding.toolbarRight.setText(text);
        mTitleViewBinding.toolbarRight.setOnClickListener(listener);
    }

    protected void setToolbarRight(String text, int color, View.OnClickListener listener) {
        mTitleViewBinding.toolbarRight.setTextColor(color);
        setToolbarRight(text, listener);
    }

    protected void setToolbarRightText(String text, int color, View.OnClickListener listener) {
        mTitleViewBinding.toolbarRight.setText(text);
        mTitleViewBinding.toolbarRight.setTextColor(color);
        mTitleViewBinding.toolbarRight.setVisibility(View.VISIBLE);
        mTitleViewBinding.toolbarRight.setOnClickListener(listener);
    }

    protected void setToolbarRightDrawable(int resId, View.OnClickListener listener) {
        mTitleViewBinding.toolbarRight.setCompoundDrawablesWithIntrinsicBounds(0, 0, resId, 0);
        mTitleViewBinding.toolbarRight.setVisibility(View.VISIBLE);
        mTitleViewBinding.toolbarRight.setOnClickListener(listener);
    }

    protected void setToolbarRightVisible(boolean isVisible) {
        mTitleViewBinding.toolbarRight.setVisibility(isVisible ? View.VISIBLE : View.INVISIBLE);
    }
}

File: WIFIDirect/app/src/main/java/com/thc/www/p2p/base/BaseFragment.kt
package com.thc.www.p2p.base

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Color
import android.os.Build
import android.os.Bundle
import android.support.v4.app.Fragment
import android.text.TextUtils
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import com.thc.www.p2p.utils.StatusBarUtils
import com.thc.www.p2p.widget.ToastUtils
import com.thc.www.wifi_direct.R
import org.greenrobot.eventbus.EventBus
import javax.inject.Inject


/**
 * @Describe
 * @Author thc
 * @Date  2018/4/20
 */
abstract class BaseFragment<P : IBasePresenter> : Fragment(), IBaseView {
    @Inject
    @JvmField
    var mPresenter: P? = null
    @Inject
    @JvmField
    var mApplication: BaseApplication? = null
    private var toolBar: View? = null // title栏

    /**
     * 重写该方法，判断是否需要添加title栏
     */
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(getLayoutId(), container, false)
        if (useToolBar()) {
            return LinearLayout(activity).apply {
                layoutParams = LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.MATCH_PARENT)
                orientation = LinearLayout.VERTICAL
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP)
                    activity!!.window.statusBarColor = resources.getColor(R.color.colorTheme)
                toolBar = inflater.inflate(R.layout.toolbar_layout, container, false)
                addView(toolBar)
                addView(view, layoutParams)
            }
        } else {
            if (fullScreen()) {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                    activity!!.window.statusBarColor = Color.TRANSPARENT
                    StatusBarUtils.setStatusTheme(activity!!.window, true)
                } else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                    activity!!.window.decorView.systemUiVisibility = View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
                    activity!!.window.statusBarColor = Color.TRANSPARENT
                }
            } else {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                    activity!!.window.statusBarColor = resources.getColor(R.color.colorTheme)
                }
            }
        }
        return view
    }

    /**
     * 判断一个一个点的灰阶
     * 灰阶小于 < 142时认为应该使用白色状态栏文字
     */
    private fun isLightStatus(bitmap: Bitmap): Boolean {
        val color = bitmap.getPixel(bitmap.width * 2 / 3, 40)
        return (Color.red(color) * 0.299 + Color.green(color) * 0.587 + Color.blue(color) * 0.114) < 142
    }

    /**
     * 重写该方法做一些初始化操作
     */
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setPresenter()
        if (useEventBus())
            EventBus.getDefault().register(this)
        initView()
        initData()
    }

    /**
     * 重写该方法，销毁资源
     */
    override fun onDestroyView() {
        super.onDestroyView()
        if (useEventBus())
            EventBus.getDefault().unregister(this)
    }

    /**
     * 设置ToolBar的title
     */
    protected fun setToolbarTitle(title: String) {
        toolBar?.findViewById<TextView>(R.id.toolbar_title)?.text = title
    }

    /**
     * 设置左侧返回按钮可用
     */
    protected fun setToolbarLeftEnable() {
        toolBar?.findViewById<ImageView>(R.id.toolbar_left)?.run {
            visibility = View.VISIBLE
            setImageBitmap(BitmapFactory.decodeResource(resources, R.mipmap.icon_back))
            setOnClickListener({ activity!!.onBackPressed() })
        }
    }

    /**
     * 设置左侧返回按钮不可用
     */
    protected fun setToolbarLeftDisEnable() {
        toolBar?.findViewById<ImageView>(R.id.toolbar_left)?.visibility = View.INVISIBLE
    }

    /**
     * 设置右侧返回按钮不可用
     */
    protected fun setToolbarRightDisEnable() {
        toolBar?.findViewById<TextView>(R.id.toolbar_right)?.visibility = View.INVISIBLE
    }

    /**
     * 设置右侧返回按钮可用，根据传入的参数显示对应的东西
     */
    protected fun setToolbarRight(resId: Int = 0, text: String = "", color: Int = 0, listener: View.OnClickListener? = null) {
        toolBar?.findViewById<TextView>(R.id.toolbar_right)?.run {
            if (resId != 0)
                setCompoundDrawablesWithIntrinsicBounds(0, 0, resId, 0)
            if (!TextUtils.isEmpty(text))
                this.text = text
            if (color != 0)
                setTextColor(color)
            if (listener != null)
                setOnClickListener(listener)
        }
    }

    protected fun setToolbarRightVisible(isVisible: Boolean) {
        toolBar?.findViewById<TextView>(R.id.toolbar_right)?.visibility =
                if (isVisible) View.VISIBLE else View.INVISIBLE
    }

    override fun onDestroy() {
        super.onDestroy()
        mPresenter?.onDestroy()
    }

    /**
     * 是否显示title栏
     */
    protected open fun useToolBar() = false

    /**
     * 是否全屏显示
     */
    protected open fun fullScreen() = false

    /**
     * 是否使用EventBus
     */
    protected open fun useEventBus() = false

    /**
     * 初始化View
     */
    protected open fun initView() {

    }

    /**
     * 初始化数据
     */
    protected open fun initData() {

    }

    /**
     * 展示加载成功的消息
     */
    override fun showSuccessMessage(message: String) {
        ToastUtils.showMessage(context, message)
    }

    /**
     * 展示加载失败的消息
     */
    override fun showFailedMessage(message: String) {
        ToastUtils.showMessage(context, message)
    }

    /**
     * 展示加载中视图
     */
    override fun showLoading(msg: String) {
    }

    /**
     * 展示加载中视图
     */
    override fun showLoading() {
    }

    /**
     * 隐藏加载中视图
     */
    override fun hideLoading() {
    }

    /**
     * 网络错误时调用
     */
    override fun onNetError() {
    }

    /**
     * 没有数据时显示空布局
     */
    override fun showEmpty() {
    }

    /**
     * 关闭activity
     */
    override fun finishActivity() {
        activity?.finish()
    }

    /**
     * 设置mPresenter
     */
    protected abstract fun setPresenter()

    /**
     * 设置布局layout
     */
    abstract fun getLayoutId(): Int
}